### ISSUE TITLE
Find a solution for removed a.out

### ISSUE THREAD

**Issue by mikrosk** [freemint/m68k-atari-mint-binutils-gdb#1]
While it seems `a.out` support somewhat survived after [this proposal](https://sourceware.org/ml/binutils/2016-09/msg00184.html) in bfd & friends, `gdb` got rid off it for good here: https://github.com/freemint/m68k-atari-mint-binutils-gdb/commit/1736a7bd96e8927c3f889a35f9153df4fd19d833.

This kills our hopes of `gdb` ever getting fixed for the mint target.

Leaving out the obvious option (transition to ELF in FreeMiNT) maybe some hybrid solution like http://d-bug.mooo.com/beyondbrown/post/gcc-6/ can be invented also for gdb (i.e. some kind of ELF loader, which makes `gdb` happy).

**Comment by vinriviere** [comment]
A solution could be to link TOS/MiNT software to ELF executable, then load that ELF executable in GDB.

Of course GDB knows how to load ELF binaries. Just in case that could help to glue that to TOS/MiNT, @rdaum has proposed an experimental ELF loader for EmuTOS ([see there](https://sourceforge.net/p/emutos/mailman/message/35630512/)).

IMHO the right solution (maybe not so hard) would be to create a new hybrid executable format: ELF with TOS header. TOS would load it as usual, while ELF tools such as GDB could load it with minor changes.

**Comment by vinriviere** [comment]
Just in case: you must understand that the MiNT patches for binutils and GDB are on separate branches, because they were created from source releases where binutils and GDB are separated.
Of course, nowadays binutils and GDB patches could be unified.

BTW, most of the MiNT patch is inside the BFD library, this is why it can benefit to both of binutils and GDB.

**Comment by rdaum** [comment]
Unfortunately Elf32 and the CP/M68k-ish "pgmld01" both have a mandatory
header in 0-position so I'm not sure how to make it happen. In ELF's case
it needs to have a magic # there, too. I think GDB would just spit its
tongue out at anything with a pgmld01 header, even if there was Elf32
content elsewhere in it somehow.

I do have a working patch to make EmuTOS execute Elf32 m68k binaries, as
long as all the relocation and symtab sections are linked in. But this
makes for quite large binaries so only makes sense for largre memory
systems. On a systems with 030 or ColdFire MMU maybe we could hack MiNT up
enough to avoid needing relocations and just load the program at a virtual
address based on the Elf header, I dunno?

I am still waiting on my employer's permission to do outside contributions
to EmuTOS. Otherwise they (Google) own the copyright, which makes
contributing the patch tricky.

It could also be done as a MiNT extension or as a separate TSR, which might
make it more useful to people, and I did start working on a TSR to do it,
but I got stuck figuring out the assembly to install in the trap. I could
revisit that, as in that case as it's a separate project there's less
paperwork involved, I can just open source it as a separate project on
GitHub under Google's copyright.

Ryan

**Comment by mikrosk** [comment]
@vinriviere I totally like the former idea, make `gdb` load the ELF binary and ... that's it. Especially for the reasons @rdaum describes (early stage, impossible to make it both formats, big memory/storage footprint, ...).

Realistically, if `gdb` understands ELF and m68k, all the calls to the OS are already within the binary ... what needs to be done?

**Comment by vinriviere** [comment]
I'm not sure if GDB can load ELF executables from itself, or if if needs support from the OS. In the latter case, @rdaum's work will be very helpful. After that, some glue needs to be added so GDB can drive the debugged process. This had already done in old GDB using MiNT's ptrace() call. And voila.

**Comment by rdaum** [comment]
Pretty sure gdb uses libelf

**Comment by mikrosk** [comment]
Just to write this down somewhere, the latest gdb 8.0 [requires C++11 compiler](http://lists.gnu.org/archive/html/info-gnu/2017-06/msg00002.html) to build. I've done some proof-of-concept checking with gcc 6.2, it seems many things that 'the brown guys' mention here: http://d-bug.mooo.com/beyondbrown/post/gcc-6/ is not really an issue, i.e. one should be able to compile gcc 6.2 without much trouble. Latest atari binutils is a proof of that.

**Comment by vinriviere** [comment]
@th-otto has already ported the MiNT patch to GCC 7.1.0. See attachment.
https://sourceforge.net/p/emutos/mailman/message/35838601/

**Comment by mikrosk** [comment]
@th-otto: :thumbsup:

Will give it a try.

**Comment by mikrosk** [comment]
The day I load an "almost-ELF" binary to an "almost-ELF" gdb and will work, I'm closing this task. :)

**Comment by th-otto** [comment]
> The day I load an "almost-ELF" binary to an "almost-ELF" gdb and will work, I'm closing this task. :)

I think that day has come ;)

**Comment by vinriviere** [comment]
Hehe... We still need a few adjustments, and proper releases. But the proof of concept is done 😎 PRG/ELF is a solid replacement for the venerable a.out format. This should be enough for the upcoming decade. Even more.

Note that PRG/ELF is very simple, almost-ELF. So it could be used with any GCC version, past or future. Not only the latest GCC 13.
